### PR TITLE
Week 3

### PR DESCRIPTION
**Pull request by kgabinete08**
Week 3
- My queue can reorder and rate videos
- feature specs
- refactor tests with macros and shared examples

### CODE REVIEW

File: app/controllers/queue_items_controller.rb [hunk 1]
```diff
@@ -11,9 +11,20 @@ def create
...
+      current_user.normalize_queue_item_positions
+    rescue ActiveRecord::RecordInvalid
+      flash[:danger] = "Invalid position number."
+    end
+    redirect_to my_queue_path
+  end
+
   def destroy
     queue_item = QueueItem.find(params[:id])
     queue_item.destroy if current_user.queue_items.include?(queue_item)
```
Comment (tomtomecek) [inline app/controllers/queue_items_controller.rb]: This part should be refactored to model - `current_user.queue_items.include?(queue_item)`

File: app/controllers/queue_items_controller.rb [hunk 2]
```diff
@@ -26,4 +37,13 @@ def queue_video(video)
   def new_queue_item_position
     current_user.queue_items.count + 1
   end
+
+  def update_queue_items
+    ActiveRecord::Base.transaction do
+      params[:queue_items].each do |queue_item_data|
+        queue_item = QueueItem.find(queue_item_data["id"])
+        queue_item.update!(position: queue_item_data["position"], rating: queue_item_data["rating"]) if queue_item.user == current_user
```
Comment (tomtomecek) [inline app/controllers/queue_items_controller.rb:45]: Good job using `.update!`

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 module ApplicationHelper
-  def stars_for_rating
-    [5,4,3,2,1].map { |number| [pluralize(number, "Star"), number] }
+  def options_for_rating(selected=nil)
```
Comment (tomtomecek) [inline app/helpers/application_helper.rb]: There should be spaces around `=`. Or any operator.

File: spec/features/user_interacts_with_queue.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
+require 'spec_helper'
+
+feature 'User interacts with the queue' do
+  scenario 'user adds and reorders videos in the queue' do
+    action = Fabricate(:category)
+    mad_max = Fabricate(:video, title: "Mad Max", category: action)
+    scarface = Fabricate(:video, title: "Scarface", category: action)
+    transporter = Fabricate(:video, title: "Transporter", category: action)
```
Comment (tomtomecek) [inline spec/features/user_interacts_with_queue.rb:8]: I would suggest to format these definitions like:

``` ruby
action      = Fabricate(:category) 
mad_max     = Fabricate(:video, title: "Mad Max", category: action) 
scarface    = Fabricate(:video, title: "Scarface", category: action) 
transporter = Fabricate(:video, title: "Transporter", category: action)
```

Slightly reads better.

File: spec/support/macros.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+def set_current_user(user=nil)
+  session[:user_id] = (user || Fabricate(:user)).id
+end
+
+def sign_in(a_user=nil)
```
Comment (tomtomecek) [inline spec/support/macros.rb]: Here you can take advantage of memoization:

``` ruby
def sign_in(user = nil)
  user ||= Fabricate(:user)
  visit sign_in_path
  fill_in 'Email Address', with: user.email
  fill_in 'Password', with: user.password
  click_button "Sign in"
end
```

File: spec/support/macros.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+def set_current_user(user=nil)
```
Comment (tomtomecek) [inline spec/support/macros.rb]: Missing spaces around `=`.

**Comment by kgabinete08** [comment]
Thanks Tomas, made some changes based on your comments.  Will merge branch now.

**Pull request merged**